lib/fileset.nix

```diff
@@ -0,0 +1,993 @@
+
+      normalised = _normaliseBase "unions" annotated;
+
+      tree = foldl' _unionTree (head normalised.trees) (tail normalised.trees);
+    in
+    if ! isList list then
+      throw "lib.fileset.unions: Expected argument to be a list, but got a ${typeOf list}."
+    else if length list == 0 then
+      throw "lib.fileset.unions: Expected argument to be a list with at least one element, but it contains no elements."
+    else
+      _create normalised.commonBase tree;
+
+  /*
+  The file set containing all files that are in both given sets.
+  Recursively, the first argument is evaluated first, only evaluating the second argument if necessary.
+
+      intersect a b == a ⋂ b
+
+  Type:
+    intersect :: FileSet -> FileSet -> FileSet
+  */
+  intersect = lhs: rhs:
+    let
+      normalised = _normaliseBase "intersect" [
+        {
+          context = "first argument";
+          value = lhs;
+        }
+        {
+          context = "second argument";
+          value = rhs;
+        }
+      ];
+    in
+    _create normalised.commonBase
+      (_intersectTree
+        (elemAt normalised.trees 0)
+        (elemAt normalised.trees 1)
+      );
+
+  /*
+  The file set containing all files that are in all the given sets.
+  Recursively, the elements are evaluated from left to right, only evaluating arguments on the right if necessary.
+
+  Type:
+    intersects :: [FileSet] -> FileSet
+  */
+  intersects = list:
+    let
+      annotated = imap0 (i: el: {
+        context = "element ${toString i} of the argument";
+        value = el;
+      }) list;
+
+      normalised = _normaliseBase "intersects" annotated;
+
+      tree = foldl' _intersectTree (head normalised.trees) (tail normalised.trees);
+    in
+    if ! isList list then
+      throw "lib.fileset.intersects: Expected argument to be a list, but got a ${typeOf list}."
+    else if length list == 0 then
+      throw "lib.fileset.intersects: Expected argument to be a list with at least one element, but it contains no elements."
+    else
+      _create normalised.commonBase tree;
+
+  /*
+  The file set containing all files that are in the first file set but not in the second.
+  Recursively, the second argument is evaluated first, only evaluating the first argument if necessary.
+
+      difference a b == a ∖ b
+
+  Type:
+    difference :: FileSet -> FileSet -> FileSet
+  */
+  difference = lhs: rhs:
+    let
+      normalised = _normaliseBase "difference" [
+        {
+          context = "first argument";
+          value = lhs;
+        }
+        {
+          context = "second argument";
+          value = rhs;
+        }
+      ];
+    in
+    _create normalised.commonBase
+      (_differenceTree normalised.commonBase
+        (elemAt normalised.trees 0)
+        (elemAt normalised.trees 1)
+      );
+
+  /*
+  Filter a file set to only contain files matching some predicate.
+
+  The predicate is called with an attribute set containing these attributes:
+
+  - `name`: The filename
```

Comment:
Not 100% clear. Is this like `baseNameOf`?

Reply:
Ah yup it is!

Reply:
Btw note that this is the draft PR. The one ready for review is https://github.com/NixOS/nixpkgs/pull/245623